library/std/src/panic.rs

```diff
@@ -316,16 +316,18 @@ where
 ///
 /// # Notes
 ///
-/// Note that this function **might not catch all panics** in Rust. A panic in
-/// Rust is not always implemented via unwinding, but can be implemented by
-/// aborting the process as well. This function *only* catches unwinding panics,
-/// not those that abort the process.
-///
-/// Note that if a custom panic hook has been set, it will be invoked before
-/// the panic is caught, before unwinding.
-///
-/// Also note that unwinding into Rust code with a foreign exception (e.g.
-/// an exception thrown from C++ code) is undefined behavior.
+/// This function **might not catch all panics** in Rust. A panic in Rust is not
+/// always implemented via unwinding, but can be implemented by aborting the
+/// process as well. This function *only* catches unwinding panics, not those
+/// that abort the process.
+///
+/// If a custom panic hook has been set, it will be invoked before the panic is
+/// caught, before unwinding.
+///
+/// Although unwinding into Rust code with a foreign exception (e.g. an
+/// exception thrown from C++ code) via an appropriate ABI (e.g. `"C-unwind"`)
+/// is permitted, catching such an exception using this function is undefined
+/// behavior.
```

Comment:
```suggestion
/// Although unwinding through Rust code with a foreign exception is permitted,
/// catching such an exception using this function is undefined behavior. An
/// example of this would be calling an `extern "C-unwind"` function and
/// attempting to catch a C++ exception using `catch_unwind`. In these cases,
/// the exception must be caught by an `extern "C"` function.
```

Feel free to adjust or not use the suggestion, but it would be good to give a clear example and say how these cases should be handled instead.

Reply:
The exception must either be uncaught (terminating the process, or at least the thread) or caught within the originating language's runtime (i.e. C++ exceptions must be caught in C++). It does not matter whether the function that _catches_ the exception is invoked with `"C"` or `"C-unwind"`, since those ABIs only have different behavior if an exception actually "escapes" through them.

```suggestion
/// Although unwinding into Rust code with a foreign exception (e.g. an
/// exception thrown from C++ code) via an appropriate ABI (e.g. `"C-unwind"`)
/// is permitted, catching such an exception using this function is undefined
/// behavior. Foreign exceptions must either be caught within the same runtime
/// in which they originated or left uncaught (i.e. permitted to terminate the
/// thread).
```

...which brings me to a point of uncertainty. @rust-lang/wg-ffi-unwind, do any of you know what happens (as currently implemented) when a foreign exception reaches the top of a thread being waited on with `std::join`? We should add that to the `std::join` documentation.

Reply:
I believe `JoinHandle`, like `main`, uses `catch_unwind`.

Reply:
Hm... so does that mean that if a C++ unwind reaches a Rust `JoinHandle` or Rust `main`, we have UB?

Reply:
It does, yes.

Reply:
@chorman0773 @tgross35 I've rewritten this bit following a [discussion in Zulip](https://rust-lang.zulipchat.com/#narrow/stream/210922-project-ffi-unwind/topic/Allowing.20catch_unwind.20for.20foreign.20exceptions/near/456229020).

Reply:
...okay, NOW I've updated this MR.